# Ensemble Learning

## Initial Imports

In [1]:
import warnings
warnings.filterwarnings('ignore')

In [2]:
import numpy as np
import pandas as pd
from pathlib import Path
from collections import Counter

In [3]:
from sklearn.metrics import balanced_accuracy_score
from sklearn.metrics import confusion_matrix
from imblearn.metrics import classification_report_imbalanced

## Read the CSV and Perform Basic Data Cleaning

In [4]:
# Load the data
file_path = Path('LoanStats_2019Q1.csv')
df = pd.read_csv(file_path)

# Preview the data
df.head()

loan_amnt  int_rate  installment home_ownership  annual_inc  \
0    10500.0    0.1719       375.35           RENT     66000.0   
1    25000.0    0.2000       929.09       MORTGAGE    105000.0   
2    20000.0    0.2000       529.88       MORTGAGE     56000.0   
3    10000.0    0.1640       353.55           RENT     92000.0   
4    22000.0    0.1474       520.39       MORTGAGE     52000.0   

  verification_status   issue_d loan_status pymnt_plan    dti  ...  \
0     Source Verified  Mar-2019    low_risk          n  27.24  ...   
1            Verified  Mar-2019    low_risk          n  20.23  ...   
2            Verified  Mar-2019    low_risk          n  24.26  ...   
3            Verified  Mar-2019    low_risk          n  31.44  ...   
4        Not Verified  Mar-2019    low_risk          n  18.76  ...   

   pct_tl_nvr_dlq  percent_bc_gt_75  pub_rec_bankruptcies  tax_liens  \
0            85.7             100.0                   0.0        0.0   
1            91.2              50.0                   1.0        0.0   
2            66.7              50.0                   0.0        0.0   
3           100.0              50.0                   1.0        0.0   
4           100.0               0.0                   0.0        0.0   

   tot_hi_cred_lim  total_bal_ex_mort total_bc_limit  \
0          65687.0            38199.0         2000.0   
1         271427.0            60641.0        41200.0   
2          60644.0            45684.0         7500.0   
3          99506.0            68784.0        19700.0   
4         219750.0            25919.0        27600.0   

   total_il_high_credit_limit  hardship_flag  debt_settlement_flag  
0                     61987.0              N                     N  
1                     49197.0              N                     N  
2                     43144.0              N                     N  
3                     76506.0              N                     N  
4                     20000.0              N                     N  

[5 rows x 86 columns]

In [5]:
from sklearn.preprocessing import LabelEncoder, StandardScaler

df=df.drop(columns= ['hardship_flag', 'debt_settlement_flag', 'pymnt_plan', 'initial_list_status', 'verification_status', 'issue_d', 'next_pymnt_d', 'application_type'])

le = LabelEncoder()

# Encoding homeowner column
le.fit(df["home_ownership"])
df["home_ownership"] = le.transform(df["home_ownership"])

df.head()




loan_amnt  int_rate  installment  home_ownership  annual_inc loan_status  \
0    10500.0    0.1719       375.35               3     66000.0    low_risk   
1    25000.0    0.2000       929.09               1    105000.0    low_risk   
2    20000.0    0.2000       529.88               1     56000.0    low_risk   
3    10000.0    0.1640       353.55               3     92000.0    low_risk   
4    22000.0    0.1474       520.39               1     52000.0    low_risk   

     dti  delinq_2yrs  inq_last_6mths  open_acc  ...  num_tl_90g_dpd_24m  \
0  27.24          0.0             0.0       8.0  ...                 0.0   
1  20.23          0.0             0.0      17.0  ...                 0.0   
2  24.26          0.0             0.0       8.0  ...                 0.0   
3  31.44          0.0             1.0      10.0  ...                 0.0   
4  18.76          0.0             1.0      14.0  ...                 0.0   

   num_tl_op_past_12m  pct_tl_nvr_dlq  percent_bc_gt_75  pub_rec_bankruptcies  \
0                 3.0            85.7             100.0                   0.0   
1                 6.0            91.2              50.0                   1.0   
2                 4.0            66.7              50.0                   0.0   
3                 1.0           100.0              50.0                   1.0   
4                 1.0           100.0               0.0                   0.0   

   tax_liens  tot_hi_cred_lim  total_bal_ex_mort  total_bc_limit  \
0        0.0          65687.0            38199.0          2000.0   
1        0.0         271427.0            60641.0         41200.0   
2        0.0          60644.0            45684.0          7500.0   
3        0.0          99506.0            68784.0         19700.0   
4        0.0         219750.0            25919.0         27600.0   

   total_il_high_credit_limit  
0                     61987.0  
1                     49197.0  
2                     43144.0  
3                     76506.0  
4                     20000.0  

[5 rows x 78 columns]

In [6]:
pd.set_option('display.max_columns', None)
df.head()

loan_amnt  int_rate  installment  home_ownership  annual_inc loan_status  \
0    10500.0    0.1719       375.35               3     66000.0    low_risk   
1    25000.0    0.2000       929.09               1    105000.0    low_risk   
2    20000.0    0.2000       529.88               1     56000.0    low_risk   
3    10000.0    0.1640       353.55               3     92000.0    low_risk   
4    22000.0    0.1474       520.39               1     52000.0    low_risk   

     dti  delinq_2yrs  inq_last_6mths  open_acc  pub_rec  revol_bal  \
0  27.24          0.0             0.0       8.0      0.0     1609.0   
1  20.23          0.0             0.0      17.0      1.0    18368.0   
2  24.26          0.0             0.0       8.0      0.0    13247.0   
3  31.44          0.0             1.0      10.0      1.0    17996.0   
4  18.76          0.0             1.0      14.0      0.0     9091.0   

   total_acc  out_prncp  out_prncp_inv  total_pymnt  total_pymnt_inv  \
0       14.0   10204.87       10204.87       355.29           355.29   
1       34.0   24293.13       24293.13       873.53           873.53   
2       21.0   19636.79       19636.79       485.44           485.44   
3       23.0    9719.34        9719.34       335.33           335.33   
4       28.0   21614.73       21614.73       484.36           484.36   

   total_rec_prncp  total_rec_int  total_rec_late_fee  recoveries  \
0           295.13          60.16                 0.0         0.0   
1           706.87         166.66                 0.0         0.0   
2           363.21         122.23                 0.0         0.0   
3           280.66          54.67                 0.0         0.0   
4           385.27          99.09                 0.0         0.0   

   collection_recovery_fee  last_pymnt_amnt  collections_12_mths_ex_med  \
0                      0.0           375.35                         0.0   
1                      0.0           929.09                         0.0   
2                      0.0           529.88                         0.0   
3                      0.0           353.55                         0.0   
4                      0.0           520.39                         0.0   

   policy_code  acc_now_delinq  tot_coll_amt  tot_cur_bal  open_acc_6m  \
0          1.0             0.0         178.0      38199.0          3.0   
1          1.0             0.0         165.0     228119.0          3.0   
2          1.0             0.0           0.0      45684.0          2.0   
3          1.0             0.0           0.0      68784.0          0.0   
4          1.0             0.0           0.0     177745.0          0.0   

   open_act_il  open_il_12m  open_il_24m  mths_since_rcnt_il  total_bal_il  \
0          4.0          2.0          4.0                 3.0       36590.0   
1          3.0          2.0          3.0                 4.0       42273.0   
2          3.0          2.0          3.0                 4.0       32437.0   
3          4.0          1.0          3.0                 9.0       50788.0   
4          1.0          0.0          2.0                14.0       16828.0   

   il_util  open_rv_12m  open_rv_24m  max_bal_bc  all_util  total_rev_hi_lim  \
0     59.0          1.0          2.0      1553.0      58.0            3700.0   
1     86.0          4.0          8.0      5197.0      61.0           50400.0   
2     75.0          2.0          3.0      2810.0      75.0           17500.0   
3     66.0          0.0          1.0      6360.0      69.0           23000.0   
4     84.0          1.0          1.0      3600.0      41.0           43100.0   

   inq_fi  total_cu_tl  inq_last_12m  acc_open_past_24mths  avg_cur_bal  \
0     1.0          0.0           4.0                   6.0       4775.0   
1     4.0          9.0           3.0                  13.0      13419.0   
2     0.0          4.0           1.0                   6.0       5711.0   
3     3.0          5.0           4.0                   4.0       6878.0   
4     2.0          0.0          

## Split the Data into Training and Testing

In [7]:
# Create our features
X = df.drop(columns=['loan_status'])

# Create our target
y = df['loan_status']
X

loan_amnt  int_rate  installment  home_ownership  annual_inc    dti  \
0        10500.0    0.1719       375.35               3     66000.0  27.24   
1        25000.0    0.2000       929.09               1    105000.0  20.23   
2        20000.0    0.2000       529.88               1     56000.0  24.26   
3        10000.0    0.1640       353.55               3     92000.0  31.44   
4        22000.0    0.1474       520.39               1     52000.0  18.76   
...          ...       ...          ...             ...         ...    ...   
68812    10000.0    0.1502       346.76               3     26000.0   9.60   
68813    12000.0    0.2727       368.37               3     63000.0  29.07   
68814     5000.0    0.1992       185.62               1     52000.0  14.86   
68815    40000.0    0.0646      1225.24               1    520000.0   9.96   
68816    16000.0    0.1131       350.36               1     72000.0   7.02   

       delinq_2yrs  inq_last_6mths  open_acc  pub_rec  revol_bal  total_acc  \
0              0.0             0.0       8.0      0.0     1609.0       14.0   
1              0.0             0.0      17.0      1.0    18368.0       34.0   
2              0.0             0.0       8.0      0.0    13247.0       21.0   
3              0.0             1.0      10.0      1.0    17996.0       23.0   
4              0.0             1.0      14.0      0.0     9091.0       28.0   
...            ...             ...       ...      ...        ...        ...   
68812          0.0             0.0       9.0      0.0     2684.0       20.0   
68813          0.0             0.0       8.0      0.0    13314.0       26.0   
68814          0.0             0.0       5.0      1.0     3715.0       14.0   
68815          0.0             1.0      21.0      0.0    59529.0       57.0   
68816          2.0             0.0      12.0      1.0    11882.0       39.0   

       out_prncp  out_prncp_inv  total_pymnt  total_pymnt_inv  \
0       10204.87       10204.87       355.29           355.29   
1       24293.13       24293.13       873.53           873.53   
2       19636.79       19636.79       485.44           485.44   
3        9719.34        9719.34       335.33           335.33   
4       21614.73       21614.73       484.36           484.36   
...          ...            ...          ...              ...   
68812    9326.86        9326.86      1019.42          1019.42   
68813   11706.42       11706.42      1086.93          1086.93   
68814    4666.78        4666.78       571.33           571.33   
68815   36953.94       36953.94      3661.36          3661.36   
68816   15395.66       15395.66      1041.03          1041.03   

       total_rec_prncp  total_rec_int  total_rec_late_fee  recoveries  \
0               295.13          60.16                 0.0         0.0   
1               706.87         166.66                 0.0         0.0   
2               363.21         122.23                 0.0         0.0   
3               280.66          54.67                 0.0         0.0   
4               385.27          99.09                 0.0         0.0   
...                ...            ...                 ...         ...   
68812           673.14         346.28                 0.0         0.0   
68813           293.58         793.35                 0.0         0.0   
68814           333.22         238.11                 0.0         0.0   
68815          3046.06         615.30                 0.0         0.0   
68816           604.34         436.69                 0.0         0.0   

       collection_recovery_fee  last_pymnt_amnt  collections_12_mths_ex_med  \
0                          0.0           375.35                         0.0   
1                          0.0           929.09                         0.0   
2                          0.0           529.88                         0.0   
3                          0.0           353.55                         0.0   
4                          0.0           520.39                     

In [8]:
X.describe()

loan_amnt      int_rate   installment  home_ownership    annual_inc  \
count  68817.000000  68817.000000  68817.000000    68817.000000  6.881700e+04   
mean   16677.594562      0.127718    480.652863        1.812779  8.821371e+04   
std    10277.348590      0.048130    288.062432        0.941313  1.155800e+05   
min     1000.000000      0.060000     30.890000        0.000000  4.000000e+01   
25%     9000.000000      0.088100    265.730000        1.000000  5.000000e+04   
50%    15000.000000      0.118000    404.560000        1.000000  7.300000e+04   
75%    24000.000000      0.155700    648.100000        3.000000  1.040000e+05   
max    40000.000000      0.308400   1676.230000        3.000000  8.797500e+06   

                dti   delinq_2yrs  inq_last_6mths      open_acc       pub_rec  \
count  68817.000000  68817.000000    68817.000000  68817.000000  68817.000000   
mean      21.778153      0.217766        0.497697     12.587340      0.126030   
std       20.199244      0.718367        0.758122      6.022869      0.336797   
min        0.000000      0.000000        0.000000      2.000000      0.000000   
25%       13.890000      0.000000        0.000000      8.000000      0.000000   
50%       19.760000      0.000000        0.000000     11.000000      0.000000   
75%       26.660000      0.000000        1.000000     16.000000      0.000000   
max      999.000000     18.000000        5.000000     72.000000      4.000000   

           revol_bal     total_acc     out_prncp  out_prncp_inv   total_pymnt  \
count   68817.000000  68817.000000  68817.000000   68817.000000  68817.000000   
mean    17604.142828     24.666347  16018.845833   16015.919519    970.639473   
std     21835.880400     12.318917  10002.695495   10004.985940   1229.563301   
min         0.000000      3.000000      0.000000       0.000000      0.000000   
25%      6293.000000     16.000000   8578.800000    8576.500000    403.760000   
50%     12068.000000     23.000000  14067.750000   14061.030000    710.790000   
75%     21735.000000     31.000000  22860.700000   22860.490000   1214.600000   
max    587191.000000    164.000000  40000.000000   40000.000000  41418.981165   

       total_pymnt_inv  total_rec_prncp  total_rec_int  total_rec_late_fee  \
count     68817.000000     68817.000000   68817.000000        68817.000000   
mean        970.431151       658.748750     311.855687            0.035072   
std        1229.569864      1126.237377     279.786954            1.086720   
min           0.000000         0.000000       0.000000            0.000000   
25%         403.550000       250.370000     115.440000            0.000000   
50%         710.790000       449.890000     225.250000            0.000000   
75%        1214.590000       762.170000     415.710000            0.000000   
max       41418.980000     40000.000000    2735.400000           75.240000   

       recoveries  collection_recovery_fee  last_pymnt_amnt  \
count     68817.0                  68817.0     68817.000000   
mean          0.0                      0.0       517.202218   
std           0.0                      0.0       816.021053   
min           0.0                      0.0         0.000000   
25%           0.0                      0.0       267.110000   
50%           0.0                      0.0       408.520000   
75%           0.0                      0.0       652.580000   
max           0.0                      0.0     40500.340000   

       collections_12_mths_ex_med  policy_code  acc_now_delinq   tot_coll_amt  \
count                68817.000000      68817.0         68817.0   68817.000000   
mean                     0.016188          1.0             0.0     188.664298   
std                      0.134774          0.0             0.0    1808.457195   
min                      0.000000          1.0             0.0       0.000000   
25%                      0.000000          1.0             0.0       0.000000   
50%                      0.000000          1.0             0.0 

In [9]:
# Check the balance of our target values
y.value_counts()

low_risk     68470
high_risk      347
Name: loan_status, dtype: int64

In [10]:
# Split the X and y into X_train, X_test, y_train, y_test
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, 
                                                   y, 
                                                   random_state=1, 
                                                   stratify=y)
X_train.shape

(51612, 77)

## Data Pre-Processing

Scale the training and testing data using the `StandardScaler` from `sklearn`. Remember that when scaling the data, you only scale the features data (`X_train` and `X_testing`).

In [11]:
# Create the StandardScaler instance
scaler = StandardScaler()

In [12]:
# Fit the Standard Scaler with the training data
# When fitting scaling functions, only train on the training dataset
X_scaler = scaler.fit(X_train)


In [13]:
# Scale the training and testing data
X_train_scaled = X_scaler.transform(X_train)
X_test_scaled = X_scaler.transform(X_test)

## Ensemble Learners

In this section, you will compare two ensemble algorithms to determine which algorithm results in the best performance. You will train a Balanced Random Forest Classifier and an Easy Ensemble classifier . For each algorithm, be sure to complete the folliowing steps:

1. Train the model using the training data. 
2. Calculate the balanced accuracy score from sklearn.metrics.
3. Display the confusion matrix from sklearn.metrics.
4. Generate a classication report using the `imbalanced_classification_report` from imbalanced-learn.
5. For the Balanced Random Forest Classifier only, print the feature importance sorted in descending order (most important feature to least important) along with the feature score

Note: Use a random state of 1 for each algorithm to ensure consistency between tests

### Balanced Random Forest Classifier

In [14]:
# Resample the training data with the BalancedRandomForestClassifier
from imblearn.ensemble import BalancedRandomForestClassifier

brf = BalancedRandomForestClassifier(n_estimators=1000, random_state=1)
brf.fit(X_train_scaled, y_train)

BalancedRandomForestClassifier(n_estimators=1000, random_state=1)

In [15]:
# Calculated the balanced accuracy score
y_pred_brf = brf.predict(X_test_scaled)
bas_brf=balanced_accuracy_score(y_test, y_pred_brf)
print(bas_brf)

0.7190656336745753


In [16]:
# Display the confusion matrix
cm_brf = confusion_matrix(y_test, y_pred_brf)
cm_df_brf = pd.DataFrame(
    cm_brf, index=["Actual High Risk", "Actual Low Risk"], columns=["Predicted High Risk", "Predicted Low Risk"]
)
cm_df_brf


Predicted High Risk  Predicted Low Risk
Actual High Risk                   50                  37
Actual Low Risk                  2338               14780

In [17]:
# Print the imbalanced classification report
print(classification_report_imbalanced(y_test, y_pred_brf))

                   pre       rec       spe        f1       geo       iba       sup

  high_risk       0.02      0.57      0.86      0.04      0.70      0.48        87
   low_risk       1.00      0.86      0.57      0.93      0.70      0.51     17118

avg / total       0.99      0.86      0.58      0.92      0.70      0.51     17205



In [18]:
# List the features sorted in descending order by feature importance
importance = brf.feature_importances_
sorted(zip(brf.feature_importances_, X.columns), reverse=True)

[(0.07506146035199632, 'total_rec_prncp'),
 (0.06503298614720944, 'total_rec_int'),
 (0.05956430658785883, 'last_pymnt_amnt'),
 (0.05541339216090808, 'total_pymnt_inv'),
 (0.0520925170119819, 'total_pymnt'),
 (0.030699289939673195, 'int_rate'),
 (0.021419120634941384, 'dti'),
 (0.018970490358724248, 'max_bal_bc'),
 (0.018559009842170748, 'annual_inc'),
 (0.01737929321018198, 'installment'),
 (0.01724424861923503, 'il_util'),
 (0.016602856920196876, 'avg_cur_bal'),
 (0.016528663101664398, 'mths_since_recent_inq'),
 (0.01636631148433249, 'out_prncp'),
 (0.016312302819700274, 'revol_bal'),
 (0.016147014375881865, 'mo_sin_old_il_acct'),
 (0.016105673298610967, 'bc_util'),
 (0.015652660217925277, 'mo_sin_old_rev_tl_op'),
 (0.015580744342436833, 'tot_cur_bal'),
 (0.015507411498261476, 'out_prncp_inv'),
 (0.015229804951023246, 'tot_hi_cred_lim'),
 (0.014889830379514777, 'total_bc_limit'),
 (0.014832402952380041, 'bc_open_to_buy'),
 (0.014579141163294698, 'all_util'),
 (0.014481453780667794, '

### Easy Ensemble Classifier

In [19]:
# Train the Classifier
from imblearn.ensemble import EasyEnsembleClassifier
eec = EasyEnsembleClassifier(n_estimators=1000, random_state=1)
eec.fit(X_train_scaled, y_train)

EasyEnsembleClassifier(n_estimators=1000, random_state=1)

In [20]:
# Calculated the balanced accuracy score
y_pred_eec = eec.predict(X_test_scaled)
bas_eec=balanced_accuracy_score(y_test, y_pred_brf)
print(bas_eec)

0.7190656336745753


In [21]:
# Display the confusion matrix
cm_eec = confusion_matrix(y_test, y_pred_eec)
cm_df_eec = pd.DataFrame(
    cm_eec, index=["Actual High Risk", "Actual Low Risk"], columns=["Predicted High Risk", "Predicted Low Risk"]
)
cm_df_eec

Predicted High Risk  Predicted Low Risk
Actual High Risk                   60                  27
Actual Low Risk                  3636               13482

In [24]:
# Print the imbalanced classification report
print(classification_report_imbalanced(y_test, y_pred_eec))

                   pre       rec       spe        f1       geo       iba       sup

  high_risk       0.02      0.69      0.79      0.03      0.74      0.54        87
   low_risk       1.00      0.79      0.69      0.88      0.74      0.55     17118

avg / total       0.99      0.79      0.69      0.88      0.74      0.55     17205



### Final Questions

1. Which model had the best balanced accuracy score?

    THe eec has a better index balance accuracy  score of .55

2. Which model had the best recall score?

   THe BRC had the better score with an average of .86

3. Which model had the best geometric mean score?

   The easy has a better score of .74 

4. What are the top three features?

     [(0.07506146035199632, 'total_rec_prncp'),
     (0.06503298614720944, 'total_rec_int'),
     (0.05956430658785883, 'last_pymnt_amnt'),